### PR TITLE
Fix chatter loading text flicker

### PR DESCRIPTION
**Pull request by xdu-odoo**

### CODE REVIEW

**Comment by xdu-odoo** [comment]
I forgot to check the tests and they failed, sorry I fix this ASAP

**Comment by xdu-odoo** [comment]
Green :+1:

File: addons/mail/static/src/owl/store/messaging_actions.js
Comment (alexkuhn) [inline addons/mail/static/src/owl/store/messaging_actions.js]: comment is not that useful, I would remove it

File: addons/mail/static/src/owl/store/messaging_actions.js
Comment (alexkuhn) [inline addons/mail/static/src/owl/store/messaging_actions.js]: This looks wrong.
A thread cache is initially not loaded, but then when it has 30 messages and a "load more" action is triggered, it should still be loaded but loading more.

I get conceptually this is strange that code in thread cache uses mutually exclusive booleans (a thread cache can't really be is loaded, loading, and loading more at the same time), and this could be replaced by a "state" (like a string, and it uses getter. Note that the reason it hasn't been done that way in the 1st place was to avoid typo with strings, and getters are not automatically observed by `useStore`...).

If I get this right, because `isLoaded` is set to `false`, template of thread during a loading more scenario could (and would most of the time in practice) re-render with no messages, and then display all messages. This would definitely flicker a lot, and would definitely mess up scroll bar adjustment.

File: addons/mail/static/src/owl/store/messaging_actions.js
Comment (alexkuhn) [inline addons/mail/static/src/owl/store/messaging_actions.js]: I like the idea, but I'm not sure the timeout should be managed in the store.
At the moment I lean toward components displaying "please wait..." in case thread cache is still loading after a certain amount of time.

Also note that you should handle clearing the timeout, otherwise you could have weird glitches like loading.../not loading.../ loading.../ not loading..., if you see what I meant

**Review by alexkuhn — CHANGES_REQUESTED** [review]

**Comment by xdu-odoo** [comment]
And it's green :1st_place_medal:

File: addons/mail/static/src/messaging/component/message_list/message_list.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/component/message_list/message_list.js]: this guard should no longer be necessary when https://github.com/odoo-dev/odoo/pull/230 is merged (very soon)

File: addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js]: I would move this logic in entity, because I think it's quite unsafe having asynchronous logic in component, so rule of thumb is usually move to entity in that case.
Also entity thread viewer directly maps a thread viewer component, so it shouldn't be an issue in this case.

File: addons/mail/static/src/messaging/test_utils.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/test_utils.js]: Already fixed by https://github.com/odoo-dev/odoo/pull/230

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js]: Would you use another field or handle this with `isLoading` or `isLoaded` ?

Comment (alexkuhn) [reply addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js]: `isLoading` and `isLoaded` is dependent on thread cache. Here it's about displaying loading or not on thread viewer, which is a bit different.

I think a new attribute should be introduced in thread viewer, and it should have timeout for auto-updating this attribute.

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/component/thread_viewer/thread_viewer.js]: Ok that's what I was thinking, I'm working on it, thx

File: addons/mail/static/src/messaging/component/message_list/message_list.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/component/message_list/message_list.js]: Will remove when #230 is merged, just keeping it to have green runbot

File: addons/mail/static/src/messaging/test_utils.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/test_utils.js]: Will remove when #230 is merged, just keeping it to have green runbot

Comment (seb-odoo) [reply addons/mail/static/src/messaging/test_utils.js]: The relevant part was merged now.

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/test_utils.js]: Currently rebasing and removing that parts

**Comment by xdu-odoo** [comment]
Green :)

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: unnecessary guard (`clearTimeout()` on non-timeout id does nothing)

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: only happens when viewer not linked to a thread. `thread.cache()` always return an obj.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Usually, boolean attributes prefixed with "has" affects the scope of logic, while "is" fine-tunes current scope of logic.
Here an "is" fits best. Suggest renaming to "isShowingLoading".

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: You can move attribute of timeout away from `_write`. The purpose of this method is to make such attributes observable from component, whether this comes from entity instances or classes. Classes are not observed due to being functions, which is why there are some tricks with `static _write()`.

Here it doesn't matter whether it is observed or not.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: should be just `isLoading`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: again, just `isLoading`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: `handleThreadCacheLoaded()`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: remove exceeding line

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: I prefer hard-coded value here

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Yes I know, what's your point ?

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: I really don't spot the difference but alright I'll change ..

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Don't check relation with thread cache if that's basically equivalent to checking relation with thread.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: so :

```
this._write({ hasLoader:false });
this._loaderTimeout = null;
```
?

Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: well, assignment to timeout should be next to `clearTimeout()`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Just took the condition that was in the template earlier, I indeed didn't understand why both were checked so ok I'll change

**Comment by xdu-odoo** [comment]
Gween !

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: just `clearTimeout()`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: move it closer to clearTimeouts. Maybe define a private method that simply make clearTimeout + setting to falsy value

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: remove from `_write()`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: again, just `clearTimeout()`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: alphabetical order

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: space after `if` keyword

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: `else` inline with closing `}` of preceding conditional branching scope

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: inline or multi-line, depending on length, but don't need `const`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: space after `<key>:`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: again

**Review by alexkuhn — CHANGES_REQUESTED** [review]
also rebase with `master-mail-owl`:
`_write()` has been removed and should be replaced with `Object.assign()`.

**Comment by xdu-odoo** [comment]
Green & rebased :D

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (seb-odoo) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]:
```suggestion
                this.isShowingLoading = false;
```

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: just assignment with `=` (several occurrences in this file)

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: pre-emptive return

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (seb-odoo) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]:
```suggestion
                this.isShowingLoading = false;
```

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (seb-odoo) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]:
```suggestion
                            this.isShowingLoading = threadCache.isLoading;
```

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (seb-odoo) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]:
```suggestion
                } else {
                    this.isShowingLoading = false;
                }
```

Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: something is off with this branch... don't know if it's because there is no management of the timeout or this code never runs in practice.

**Review by alexkuhn — COMMENTED** [review]
maybe some duplicated with prev seb comments

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: This code is executed when there is a thread and the threadCache is not loading. It could happen if you create a theadviewer on a thread already loaded. I don't know if it's done somewhere but maybe it's the case ?

It can also be executed if an update is triggered when the threadCache is loaded and the timeout is not yet executed (I mean, he's created and waiting 400 ms, maybe the update is triggered after 200ms). So we set the boolean to false already to not wait the timeout to do it (even if it doesn't change a lot in this situation).

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: We can indeed in this branch cancel the timeout if you think it's best ?

Comment (seb-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Yes, no need to let it run for nothing then, it might even potentially conflict if there is another change in between.

Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Creating a thread viewer on already loaded thread definitely happens, especially with chat windows.
I think the problem I have with this branch is that it shows `isShowingLoading` and its timeout are a bit desynchronized.
Maybe that's because change of thread or thread cache should make changes to both the timeout and recompute `isShowingLoading` at the same time.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: `this._loaderTimeout = setTimeout(() => this.isShowingLoading = true, 400);`

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: IMO this method should set `isShowingLoading` to `false` and remove the timeout (also rename function appropriately)

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: if you make the other changes, I think just keep these LOCs

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Mmh what if the timeout is executed when the thread is loaded ? That'll show the loading or it should not, no ?

Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Showing the loading happens when viewer detects cache is loading and it is still loading after the timeout. Changes of thread or thread cache or when cache becomes loaded, any timeout is aborted and `isShowingLoading` is unset.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: `_hideLoading` ?

Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Don't like "hide"...
Something like `_prepareLoading()`/`_stopLoading()`: former is a delayed "set loading", while second is an "unset loading".
Each of them handle `isShowingLoading` and timeout accordingly.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: There is no way the timeout is cancelled after 399s, and the loader is hidden but timeout function is still executed showing the loading ? If there is no way such thing could happen, then I agree with you else I think it could be a problem as the loading will be shown "forever" (would need a _update or a thread loaded event to trigger the removal of the loading)

Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Everything is single-threaded and a `clearTimeout()` always prevent execution of code in corresponding `setTimeout()`.

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Ok good to know :)

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: looks like this part can be dropped

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Not convinced but I'll change, if we spot a glitch with that, we can bring it back easily :)

**Comment by xdu-odoo** [comment]
Rebased & green

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (seb-odoo) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Maybe move this condition inside `_prepareLoading`.

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: A big wise master ( @alexkuhn -san) told me one day that he doesn't like function with a body starting with an `if` so I avoided it...

Comment (seb-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Let's see what he has to say about it then. The reasoning is that we moved all the technical details about this timeout into those methods, it feels strange to not go all the way.

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Yep, this makes sense to me too

Comment (xdu-odoo) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: The final word is ? :D

Comment (alexkuhn) [reply addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: Prevent leak of timeout:

1. either by changing the conditional expression (e.g. `!this.isShowingLoading && !this.isPreparingLoading`);
2. or move it inside `_prepareLoading` but then rename this function so that it makes sense to do nothing if a loading is already shown or in preparation.

I prefer 1st solution because `_prepareLoading()` is already smart enough from POV of users of this function, so it becomes tough to find a clean naming for this function if we move the `if` statement in.

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (seb-odoo) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js:161]: `isPreparingLoading` must be set false here too

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: prefix with `_` (private)

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: include nested `if` conditional expression in top-level `if`.

**Review by alexkuhn — CHANGES_REQUESTED** [review]
hopefully last requested changes before merge 🍡

File: addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js
Comment (alexkuhn) [inline addons/mail/static/src/messaging/entity/thread_viewer/thread_viewer.js]: syntax error (exceeding `}`) + move `)` to newline

**Review by alexkuhn — APPROVED** [review]

**Comment by seb-odoo** [comment]
Squash and merge then! (I still haven't tested but I'll complain later if this is not enough)

**Pull request merged**
